Add CI tasks for creating commit statuses on GitHub

asmith-mdsol opened the pull request:
This change uses the GitHub commit status API to set the status of the latest commit dependent on the outcome of the `ci` task. Usage notes included in the README update.

Any of @jcarres-mdsol, @mszenher or @chad-medi to review, please.

asmith-mdsol committed:
Remove duplicate clean call from ci task

The `ensure` rather than `rescue` makes the clean task run regardless of
whether the config and run tasks failed or not.

asmith-mdsol committed:
Add ci:status:(pending|success|failure) tasks

asmith-mdsol committed:
Link status tasks to main ci task

asmith-mdsol committed:
Add GitHub commit status notes to README

asmith-mdsol committed:
Tidy up owners section in README

jcarres-mdsol commented on lib/dice_bag/tasks/ci.rake:
```diff
@@ -1,9 +1,13 @@
+
 desc "Configure and run continuous integration tests then clean up"
-task :ci do 
+task :ci => ['ci:status:pending'] do
```
I like running ci task in my local machine also.  So I will be hitting this every time. It is kind of ok because I will get the "skipping setting ..."  error I guess

asmith-mdsol replied:
That was my intention. I was this to be as configuration free as possible. Also, you could just `rake ci:config ci:run`.

jcarres-mdsol commented on README.md:
````diff
@@ -41,6 +41,35 @@ namespaces:
+Create an authorization token with the following command:
+
+```
+curl -u <username> -d '{"scopes":["repo:status"],"note":"CI status updater"}' https://api.github.com/authorizations
````
very nice documented!

jcarres-mdsol commented on lib/dice_bag/github.rb:
```diff
@@ -0,0 +1,53 @@
+
+      body = %Q({
+        "state": "#{state}",
+        "target_url": "#{config.build_url}",
```
where does config.build_url come from? Jenkins? Doesn't that make us dependent of it?

asmith-mdsol replied:
It comes from the environment variable `BUILD_URL`. Jenkins does set this up for us but different build systems could easily set this too on the command line with `rake BUILD_URL=http://... ci` or rename other environment varaibles with `rake BUILD_URL=$MY_BUILD_URL ci`.

asmith-mdsol replied:
I'm adding some further documentation to explain this.

jcarres-mdsol commented:
+1 
Let's merge it

jcarres-mdsol commented on lib/dice_bag/tasks/ci.rake:
```diff
@@ -35,5 +39,20 @@ namespace :ci do
+    config = DiceBag::Configuration.new
+
+    task :pending do
+      DiceBag::GitHub.update_commit_status('pending', config)
```
yes, this is what github expects. 
making these constants is generally a best practice but will be kind of pointless 
PENDING = 'pending' 
SUCCESS  = 'success'
etc.

asmith-mdsol replied:
I'll switch these for symbols and that'll become the interface to the method. I'll just `to_s` internally but the caller doesn't need to know that!

asmith-mdsol committed:
Add opinionated .rvmrc file

asmith-mdsol committed:
Make update_commit_status state param symbol-based

This is to better encapsulate the state values required by the GitHub
API in the method call. Also adds a full description of the method.

asmith-mdsol committed:
Add README notes on setting build number and URL

asmith-mdsol commented:
@jcarres-mdsol @chad-medi I've addressed all the comments. Ready for merge if you are!

jcarres-mdsol commented:
Very good PR

asmith-mdsol commented:
No particular reason for 1.9.2 other than I don't have 1.9.3 built locally yet and most everything else is still on 1.9.2.

jcarres-mdsol commented:
Is fine, I'll wait you here in 1.9.3, come when you want  :p
